tsdb/engine/tsm1/cache.go

```diff
@@ -0,0 +1,320 @@
+package tsm1
+
+import (
+	"container/list"
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/influxdb/influxdb/tsdb"
+)
+
+var ErrCacheMemoryExceeded = fmt.Errorf("cache maximum memory size exceeded")
+var ErrCacheInvalidCheckpoint = fmt.Errorf("invalid checkpoint")
+
+// lru orders string keys from least-recently used to most-recently used. It is not
+// goroutine safe.
+type lru struct {
+	list     *list.List
+	elements map[string]*list.Element
+}
+
+// newLRU returns an initialized LRU.
+func newLRU() *lru {
+	return &lru{
+		list:     list.New(),
+		elements: make(map[string]*list.Element),
+	}
+}
+
+// MoveToFront marks key as the most recently used key.
+func (l *lru) MoveToFront(key string) {
+	e, ok := l.elements[key]
+	if !ok {
+		l.elements[key] = l.list.PushFront(key)
+		return
+	}
+	l.list.MoveToFront(e)
+}
+
+// Remove removes key from the LRU. If the key does not exist nothing happens.
+func (l *lru) Remove(key string) {
+	if _, ok := l.elements[key]; ok {
+		l.list.Remove(l.elements[key])
+		delete(l.elements, key)
+	}
+}
+
+// Front returns the most-recently used key. If there is no such key, then "" is returned.
+func (l *lru) Front() string {
+	e := l.list.Front()
+	if e == nil {
+		return ""
+	}
+	return e.Value.(string)
+}
+
+// Back returns the least-recently used key. If there is no such key, then "" is returned.
+func (l *lru) Back() string {
+	e := l.list.Back()
+	if e == nil {
+		return ""
+	}
+	return e.Value.(string)
+}
+
+// DoFromLeast iterates through the LRU, from least-recently used to most-recently used,
+// calling the given function with each key.
+func (l *lru) DoFromLeast(f func(key string)) {
+	for e := l.list.Back(); e != nil; e = e.Prev() {
+		f(e.Value.(string))
+	}
+}
+
+// entry is the set of all values received for a given key.
+type entry struct {
+	values   Values // All stored values.
+	unsorted bool   // Whether the data requires sorting and deduping before query.
+	size     uint64 // Total Number of point-calculated bytes stored by this entry.
+}
+
+// newEntry returns a new instance of entry.
+func newEntry() *entry {
+	return &entry{}
+}
+
+// add adds the given values to the entry.
+func (e *entry) add(values []Value) {
+	for _, v := range values {
+		// Only mark unsorted if not already marked.
+		if !e.unsorted && len(e.values) > 1 {
+			e.unsorted = e.values[len(e.values)-1].Time().UnixNano() >= v.Time().UnixNano()
+		}
+		e.values = append(e.values, v)
+		e.size += uint64(v.Size())
+	}
+}
+
+// dedupe remove duplicate entries for the same timestamp and sorts the entries.
+func (e *entry) dedupe() {
+	if !e.unsorted {
+		return
+	}
+	e.values = e.values.Deduplicate()
+	e.unsorted = false
+
+	// Update size.
+	e.size = 0
+	for _, v := range e.values {
+		e.size += uint64(v.Size())
+	}
+}
+
+type entries struct {
+	ee map[uint64]*entry
```

Comment:
Yes, that is right. This is first pass at a design that doesn't require a checkpoint-per-point. In the case where large numbers of points have the same checkpoint, this means we don't pay the price of 8-bytes per point, which ~10% of a float point. It also makes eviction very easy.

Depending on how we want to solve the problem, there are, of course, other ways to lay out this data.